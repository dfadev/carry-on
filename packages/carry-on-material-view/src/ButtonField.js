import React from "react";
import Button from "@material-ui/core/Button";
import withStyles from "@material-ui/core/styles/withStyles";
import { Field } from "carry-on-react-forms";
import getFieldStatus from "./getFieldStatus";
import eventHandler from "./eventHandler";

const ButtonField = withStyles(
  {},
  { name: "CoButtonField" }
)(
  ({
    name,
    label = name,
    value: valueProp,
    disabled: disabledProp,
    onClick,
    ...props
  }) => (
    <Field path={name} type="button">
      {(field, store) => {
        const { disabled, value } = getFieldStatus(
          field,
          store,
          valueProp,
          disabledProp
        );

        return (
          <Button
            {...props}
            disabled={disabled}
            {...field.element}
            value={value}
            onClick={eventHandler(
              onClick,
              props.id || field.element.id,
              value,
              store
            )}
          >
            {label}
          </Button>
        );
      }}
    </Field>
  )
);

export default ButtonField;
