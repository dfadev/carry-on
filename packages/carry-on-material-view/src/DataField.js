import React from "react";
import withStyles from "@material-ui/core/styles/withStyles";
import { Field } from "carry-on-react-forms";

const DataField = withStyles(
  {},
  { name: "CoDataField" }
)(({ name, value, transform }) => (
  <Field path={name} readOnly>
    {(field, store) => {
      let fieldValue;
      if (value !== undefined) {
        fieldValue = typeof value === "function" ? value(field, store) : value;
      } else {
        fieldValue = field.element.value;
      }

      if (transform) fieldValue = transform(fieldValue);

      return fieldValue;
    }}
  </Field>
));

export default DataField;
