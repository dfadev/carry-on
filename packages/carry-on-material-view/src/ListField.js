import React from "react";
import AddIcon from "@material-ui/icons/Add";
import Button from "@material-ui/core/Button";
import DeleteForeverIcon from "@material-ui/icons/DeleteForever";
import IconButton from "@material-ui/core/IconButton";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import Toolbar from "@material-ui/core/Toolbar";
import withStyles from "@material-ui/core/styles/withStyles";
import { withNodesToProp } from "carry-on-react";
import { Field, FieldContext } from "carry-on-react-forms";
import { getIn, mutateSet } from "carry-on-utils";
import getFieldStatus from "./getFieldStatus";

const addRowToolbarStyles = theme => ({
  leftIcon: {
    marginRight: theme.spacing(1)
  }
});

const AddRowButton = withStyles(addRowToolbarStyles, {
  name: "CoRowListFieldAddRowButton"
})(({ classes, onClick, label, FieldAddIcon, variant, disabled }) => (
  <Toolbar>
    <Button variant={variant} onClick={onClick} disabled={disabled}>
      <FieldAddIcon fontSize="small" className={classes.leftIcon} />
      {label}
    </Button>
  </Toolbar>
));

const ListField = ({
  classes,
  name,
  add: { icon: FieldAddIcon = AddIcon, label: addLabel = "Add" } = {},
  delete: { icon: FieldDeleteIcon = DeleteForeverIcon } = {},
  max = -1,
  readOnly = false,
  defaultValues = {},
  disabled: disabledProp,
  hideRemove,
  dense,
  value: valueProp,
  renderItem
}) => (
  <Field path={name} readOnly={readOnly} type="list">
    {(field, store) => {
      const { disabled, value } = getFieldStatus(
        field,
        store,
        valueProp,
        disabledProp
      );

      if (!value) return null;

      return (
        <>
          <List dense={dense}>
            {value.map((item, idx) => {
              const p = `${name}[${idx}]`;

              return (
                <ListItem key={p} id={p}>
                  <FieldContext.Provider value={{ prefix: p }}>
                    {renderItem}
                  </FieldContext.Provider>
                  {!hideRemove && (
                    <IconButton
                      className={classes.closeMenuIcon}
                      onClick={() => {
                        store.set(state => {
                          const arr = getIn(state.values, name);
                          if (arr) arr.splice(idx, 1);
                        });
                      }}
                    >
                      <FieldDeleteIcon />
                    </IconButton>
                  )}
                </ListItem>
              );
            })}
          </List>
          {!field.readOnly && (max === -1 || max > value.length) && (
            <AddRowButton
              disabled={disabled}
              onClick={() => {
                store.set(state => {
                  let arr = getIn(state.values, name);
                  if (!arr) {
                    arr = [];
                    mutateSet(state.values, name, arr);
                  }

                  const type = typeof defaultValues;
                  if (type === "object") arr.push({ ...defaultValues });
                  else if (type === "function") arr.push(defaultValues(store));
                  else arr.push(defaultValues);
                });
              }}
              label={addLabel}
              FieldAddIcon={FieldAddIcon}
            />
          )}
        </>
      );
    }}
  </Field>
);

export default withNodesToProp("renderItem")(
  withStyles(
    {
      closeMenuIcon: {
        marginLeft: "auto"
      }
    },
    { name: "CoListField" }
  )(ListField)
);
