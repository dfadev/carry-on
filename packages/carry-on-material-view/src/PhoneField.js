import React, { useMemo } from "react";
import withStyles from "@material-ui/core/styles/withStyles";
import { AsYouType } from "libphonenumber-js";
import MaskedTextField from "./MaskedTextField";

const accept = /\d|\+/g;

const PhoneField = withStyles(
  {},
  { name: "CoPhoneField" }
)(({ defaultCountry = "US", defaultCallingCode, ...props }) => {
  const format = useMemo(() => {
    const asYouType = new AsYouType({ defaultCountry, defaultCallingCode });
    return string => {
      asYouType.reset();
      const result = asYouType.input(string);
      return result;
    };
  }, [defaultCountry, defaultCallingCode]);

  return <MaskedTextField accept={accept} format={format} {...(props || {})} />;
});

export default PhoneField;
