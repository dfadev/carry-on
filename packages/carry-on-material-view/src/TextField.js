import React from "react";
import MuiTextField from "@material-ui/core/TextField";
import withStyles from "@material-ui/core/styles/withStyles";
import { Field } from "carry-on-react-forms";
import getFieldStatus from "./getFieldStatus";

const TextField = withStyles(
  { root: { whiteSpace: "nowrap" } },
  { name: "CoTextField" }
)(
  ({
    name,
    label = name,
    disabled: disabledProp,
    value: valueProp,
    readOnly,
    InputProps,
    ...props
  }) => (
    <Field path={name} readOnly={readOnly}>
      {(field, store) => {
        const { element } = field;
        const { error, hasError, disabled, value } = getFieldStatus(
          field,
          store,
          valueProp,
          disabledProp
        );

        const realInputProps =
          typeof InputProps === "function" ? InputProps(field) : InputProps;

        return (
          <MuiTextField
            label={label}
            error={hasError}
            helperText={hasError && error}
            disabled={disabled}
            inputProps={{ ...element, value }}
            onBlur={element.onBlur}
            onChange={element.onChange}
            InputProps={realInputProps /* eslint-disable-line */}
            {...props}
          />
        );
      }}
    </Field>
  )
);

export default TextField;
