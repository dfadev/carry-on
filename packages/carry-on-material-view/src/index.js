export { default as ButtonField } from "./ButtonField";
export { default as CheckboxField } from "./CheckboxField";
export { default as CurrencyField } from "./CurrencyField";
export { default as EINField } from "./EINField";
export { default as IconButtonField } from "./IconButtonField";
export { default as MaskedTextField } from "./MaskedTextField";
export { default as PercentageField } from "./PercentageField";
export { default as PhoneField } from "./PhoneField";
export { default as RegisterMaterialView } from "./RegisterMaterialView";
export { default as ResetButtonField } from "./ResetButtonField";
export { default as SSNField } from "./SSNField";
export { default as SubmitButtonField } from "./SubmitButtonField";
export { default as TextField } from "./TextField";
export { default as ZipCodeField } from "./ZipCodeField";
export { default as getFieldStatus } from "./getFieldStatus";
export {
  default as registerComponents,
  materialViewComponents
} from "./registerComponents";
export { default as SelectField } from "./SelectField";
export { default as RadioField } from "./RadioField";
export { default as SwitchField } from "./SwitchField";
export { default as ToggleButtonGroupField } from "./ToggleButtonGroupField";
export { default as ContentField } from "./ContentField";
export { default as InspectorField } from "./InspectorField";
export { default as DateField } from "./DateField";
export { default as DobField } from "./DobField";
export { default as TimeField } from "./TimeField";
export { default as DateTimeField } from "./DateTimeField";
export { default as ListField } from "./ListField";
export { default as DataField } from "./DataField";
export { default as TabsField } from "./TabsField";
export { default as SubmitIconField } from "./SubmitIconField";
export { default as SubmitFabField } from "./SubmitFabField";
export { default as FabField } from "./FabField";
export { default as ImageField } from "./ImageField";
export { default as ButtonGroupField } from "./ButtonGroupField";
