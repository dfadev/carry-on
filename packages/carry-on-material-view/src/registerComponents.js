import React from "react";
import { register } from "carry-on-store";
import { Box, Paper, Grid } from "@mui/material";
import TextField from "./TextField";
import CheckboxField from "./CheckboxField";
import ButtonField from "./ButtonField";
import IconButtonField from "./IconButtonField";
import CurrencyField from "./CurrencyField";
import SSNField from "./SSNField";
import EINField from "./EINField";
import PercentageField from "./PercentageField";
import PhoneField from "./PhoneField";
import ZipCodeField from "./ZipCodeField";
import SubmitButtonField from "./SubmitButtonField";
import ResetButtonField from "./ResetButtonField";
import SelectField from "./SelectField";
import RadioField from "./RadioField";
import SwitchField from "./SwitchField";
import ToggleButtonGroupField from "./ToggleButtonGroupField";
import ContentField from "./ContentField";
import InspectorField from "./InspectorField";
import DateField from "./DateField";
import DobField from "./DobField";
import TimeField from "./TimeField";
import DateTimeField from "./DateTimeField";
import ListField from "./ListField";
import DataField from "./DataField";
import TabsField from "./TabsField";
import SubmitIconField from "./SubmitIconField";
import SubmitFabField from "./SubmitFabField";
import FabField from "./FabField";
import ImageField from "./ImageField";
import ButtonGroupField from "./ButtonGroupField";

function FormViewBox({ children }) {
  return (
    <Paper elevation={3}>
      <Box m={1} p={1}>
        {children}
      </Box>
    </Paper>
  );
}

function SectionBox({ section, children, ...props }) {
  return <Box {...props}>{children}</Box>;
}

function View({ children, ...props }) {
  return (
    <Grid {...props} container>
      {children}
    </Grid>
  );
}

function ViewItem({ children, field, ...props }) {
  return (
    <Grid {...props} item>
      <Box mb={1} p={1}>
        {children}
      </Box>
    </Grid>
  );
}

export const materialViewComponents = {
  components: {
    FormViewBox,
    SectionBox,
    View,
    ViewItem,
    editors: {
      button: ButtonField,
      buttonGroup: ButtonGroupField,
      checkbox: CheckboxField,
      content: ContentField,
      currency: CurrencyField,
      data: DataField,
      date: DateField,
      dateTime: DateTimeField,
      dob: DobField,
      ein: EINField,
      fab: FabField,
      iconButton: IconButtonField,
      image: ImageField,
      inspector: InspectorField,
      list: ListField,
      percentage: PercentageField,
      phone: PhoneField,
      radio: RadioField,
      reset: ResetButtonField,
      select: SelectField,
      ssn: SSNField,
      submit: SubmitButtonField,
      submitFab: SubmitFabField,
      submitIcon: SubmitIconField,
      switch: SwitchField,
      tabs: TabsField,
      text: TextField,
      time: TimeField,
      toggleButtonGroup: ToggleButtonGroupField,
      zipCode: ZipCodeField
    }
  }
};

function registerComponents(...args) {
  register({ state: materialViewComponents }, ...args);
}

export default registerComponents;
