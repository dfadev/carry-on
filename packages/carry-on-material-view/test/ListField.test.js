/**
 * @jest-environment jsdom
 */
import React from "react";
import { initStores } from "carry-on-store";
import { render } from "@testing-library/react";
import { Paper, Box } from "@material-ui/core";
import { Store, Register } from "carry-on-react";
import { Form } from "carry-on-react-forms";
import {
  FormView,
  Fields,
  Field,
  Sections,
  Section,
  InitialValues,
  FormViewer
} from "carry-on-react-view";
import { materialViewComponents } from "../src/registerComponents";
import TextField from "../src/TextField";
import PhoneField from "../src/PhoneField";
import DataField from "../src/DataField";
import ListField from "../src/ListField";

beforeEach(() => {
  initStores();
});

it("should render <ListField> inside a Form", () => {
  expect(
    render(
      <Store id={"listFieldForm"}>
        <Form>
          <InitialValues>
            {{
              contacts: [
                {
                  name: "John Johnson",
                  phone: "+44 7911 123456",
                  email: "jj@co.uk"
                },
                { name: "Jan Jansen", phone: "7025551212", email: "jj@lv.nv" }
              ]
            }}
          </InitialValues>
          <Paper>
            <Box px={3} pt={2} pb={3} mb={2}>
              <ListField name="contacts" label="Contact Editor">
                <Box my={1} p={1} pt={0}>
                  <Box>
                    Sample Greeting:
                    <h4>
                      Hello <DataField name="name" />!
                    </h4>
                  </Box>
                  <TextField name="name" label="Name" />
                  <PhoneField name="phone" label="Phone" defaultCountry="US" />
                  <TextField name="email" label="Email" />
                </Box>
              </ListField>
            </Box>
          </Paper>
        </Form>
      </Store>
    ).asFragment()
  ).toMatchSnapshot();
});

it("should render <ListField> inside a FormView", () => {
  expect(
    render(
      <Store id={"listFieldFormView"}>
        <FormView>
          <Register>{materialViewComponents}</Register>
          <InitialValues>
            {{
              contacts: [
                {
                  name: "John Johnson",
                  phone: "+44 7911 123456",
                  email: "jj@co.uk"
                },
                { name: "Jan Jansen", phone: "7025551212", email: "jj@lv.nv" }
              ]
            }}
          </InitialValues>
          <Fields>
            <Field
              name="contacts"
              label="Contact Editor"
              editor="list"
              renderItem={
                <FormViewer
                  fields={{
                    greeting: {
                      editor: "content",
                      content: (
                        <Box>
                          Sample Greeting:
                          <h4>
                            Hello <DataField name="name" />!
                          </h4>
                        </Box>
                      )
                    },
                    name: { label: "Name" },
                    phone: {
                      label: "Mobile",
                      editor: "phone",
                      defaultCountry: "US"
                    },
                    email: { label: "Email" }
                  }}
                  sections={[
                    {
                      layout: [["greeting"], ["name", "email", "phone"]]
                    }
                  ]}
                />
              }
            />
          </Fields>
          <Sections>
            <Section>{["contacts"]}</Section>
          </Sections>
        </FormView>
      </Store>
    ).asFragment()
  ).toMatchSnapshot();
});
