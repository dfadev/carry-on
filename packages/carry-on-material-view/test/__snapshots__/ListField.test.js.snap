// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render <ListField> inside a Form 1`] = `
<DocumentFragment>
  <form
    id="listFieldForm.form"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiBox-root MuiBox-root-1"
      >
        <ul
          class="MuiList-root MuiList-padding"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters"
            id="contacts[0]"
          >
            <div
              class="MuiBox-root MuiBox-root-3"
            >
              <div
                class="MuiBox-root MuiBox-root-4"
              >
                Sample Greeting:
                <h4>
                  Hello John Johnson!
                </h4>
              </div>
              <div
                class="MuiFormControl-root MuiTextField-root CoTextField-root-5"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                  data-shrink="true"
                >
                  Name
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input"
                    id="listFieldForm.form.contacts[0].name"
                    name="listFieldForm.form.contacts[0].name"
                    type="text"
                    value="John Johnson"
                  />
                </div>
              </div>
              <div
                class="MuiFormControl-root MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                  data-shrink="true"
                >
                  Phone
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input"
                    id="listFieldForm.form.contacts[0].phone"
                    name="listFieldForm.form.contacts[0].phone"
                    type="text"
                    value="+44 7911 123456"
                  />
                </div>
              </div>
              <div
                class="MuiFormControl-root MuiTextField-root CoTextField-root-5"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                  data-shrink="true"
                >
                  Email
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input"
                    id="listFieldForm.form.contacts[0].email"
                    name="listFieldForm.form.contacts[0].email"
                    type="text"
                    value="jj@co.uk"
                  />
                </div>
              </div>
            </div>
            <button
              class="MuiButtonBase-root MuiIconButton-root CoListField-closeMenuIcon-2"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </li>
          <li
            class="MuiListItem-root MuiListItem-gutters"
            id="contacts[1]"
          >
            <div
              class="MuiBox-root MuiBox-root-6"
            >
              <div
                class="MuiBox-root MuiBox-root-7"
              >
                Sample Greeting:
                <h4>
                  Hello Jan Jansen!
                </h4>
              </div>
              <div
                class="MuiFormControl-root MuiTextField-root CoTextField-root-5"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                  data-shrink="true"
                >
                  Name
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input"
                    id="listFieldForm.form.contacts[1].name"
                    name="listFieldForm.form.contacts[1].name"
                    type="text"
                    value="Jan Jansen"
                  />
                </div>
              </div>
              <div
                class="MuiFormControl-root MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                  data-shrink="true"
                >
                  Phone
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input"
                    id="listFieldForm.form.contacts[1].phone"
                    name="listFieldForm.form.contacts[1].phone"
                    type="text"
                    value="(702) 555-1212"
                  />
                </div>
              </div>
              <div
                class="MuiFormControl-root MuiTextField-root CoTextField-root-5"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                  data-shrink="true"
                >
                  Email
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input"
                    id="listFieldForm.form.contacts[1].email"
                    name="listFieldForm.form.contacts[1].email"
                    type="text"
                    value="jj@lv.nv"
                  />
                </div>
              </div>
            </div>
            <button
              class="MuiButtonBase-root MuiIconButton-root CoListField-closeMenuIcon-2"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </li>
        </ul>
        <div
          class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root CoRowListFieldAddRowButton-leftIcon-8 MuiSvgIcon-fontSizeSmall"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
              Add
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`should render <ListField> inside a FormView 1`] = `
<DocumentFragment>
  <form
    id="listFieldFormView.form"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation3 MuiPaper-rounded"
    >
      <div
        class="MuiBox-root MuiBox-root-9"
      >
        <div
          class="MuiBox-root MuiBox-root-10"
        >
          <div
            class="MuiGrid-root MuiGrid-container"
          >
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiBox-root MuiBox-root-11"
              >
                <ul
                  class="MuiList-root MuiList-padding"
                >
                  <li
                    class="MuiListItem-root MuiListItem-gutters"
                    id="contacts[0]"
                  >
                    <div
                      class="MuiPaper-root MuiPaper-elevation3 MuiPaper-rounded"
                    >
                      <div
                        class="MuiBox-root MuiBox-root-13"
                      >
                        <div
                          class="MuiBox-root MuiBox-root-14"
                        >
                          <div
                            class="MuiGrid-root MuiGrid-container"
                          >
                            <div
                              class="MuiGrid-root MuiGrid-item"
                            >
                              <div
                                class="MuiBox-root MuiBox-root-15"
                              >
                                <div
                                  class="MuiBox-root MuiBox-root-16"
                                >
                                  Sample Greeting:
                                  <h4>
                                    Hello John Johnson!
                                  </h4>
                                </div>
                              </div>
                            </div>
                          </div>
                          <div
                            class="MuiGrid-root MuiGrid-container"
                          >
                            <div
                              class="MuiGrid-root MuiGrid-item"
                            >
                              <div
                                class="MuiBox-root MuiBox-root-17"
                              >
                                <div
                                  class="MuiFormControl-root MuiTextField-root CoTextField-root-18"
                                >
                                  <label
                                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                                    data-shrink="true"
                                  >
                                    Name
                                  </label>
                                  <div
                                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                                  >
                                    <input
                                      aria-invalid="false"
                                      class="MuiInputBase-input MuiInput-input"
                                      id="listFieldFormView.form.contacts[0].name"
                                      name="listFieldFormView.form.contacts[0].name"
                                      type="text"
                                      value="John Johnson"
                                    />
                                  </div>
                                </div>
                              </div>
                            </div>
                            <div
                              class="MuiGrid-root MuiGrid-item"
                            >
                              <div
                                class="MuiBox-root MuiBox-root-19"
                              >
                                <div
                                  class="MuiFormControl-root MuiTextField-root CoTextField-root-18"
                                >
                                  <label
                                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                                    data-shrink="true"
                                  >
                                    Email
                                  </label>
                                  <div
                                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                                  >
                                    <input
                                      aria-invalid="false"
                                      class="MuiInputBase-input MuiInput-input"
                                      id="listFieldFormView.form.contacts[0].email"
                                      name="listFieldFormView.form.contacts[0].email"
                                      type="text"
                                      value="jj@co.uk"
                                    />
                                  </div>
                                </div>
                              </div>
                            </div>
                            <div
                              class="MuiGrid-root MuiGrid-item"
                            >
                              <div
                                class="MuiBox-root MuiBox-root-20"
                              >
                                <div
                                  class="MuiFormControl-root MuiTextField-root"
                                >
                                  <label
                                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                                    data-shrink="true"
                                  >
                                    Mobile
                                  </label>
                                  <div
                                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                                  >
                                    <input
                                      aria-invalid="false"
                                      class="MuiInputBase-input MuiInput-input"
                                      id="listFieldFormView.form.contacts[0].phone"
                                      name="listFieldFormView.form.contacts[0].phone"
                                      type="text"
                                      value="+44 7911 123456"
                                    />
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                    <button
                      class="MuiButtonBase-root MuiIconButton-root CoListField-closeMenuIcon-12"
                      tabindex="0"
                      type="button"
                    >
                      <span
                        class="MuiIconButton-label"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
                          />
                        </svg>
                      </span>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </li>
                  <li
                    class="MuiListItem-root MuiListItem-gutters"
                    id="contacts[1]"
                  >
                    <div
                      class="MuiPaper-root MuiPaper-elevation3 MuiPaper-rounded"
                    >
                      <div
                        class="MuiBox-root MuiBox-root-21"
                      >
                        <div
                          class="MuiBox-root MuiBox-root-22"
                        >
                          <div
                            class="MuiGrid-root MuiGrid-container"
                          >
                            <div
                              class="MuiGrid-root MuiGrid-item"
                            >
                              <div
                                class="MuiBox-root MuiBox-root-23"
                              >
                                <div
                                  class="MuiBox-root MuiBox-root-24"
                                >
                                  Sample Greeting:
                                  <h4>
                                    Hello Jan Jansen!
                                  </h4>
                                </div>
                              </div>
                            </div>
                          </div>
                          <div
                            class="MuiGrid-root MuiGrid-container"
                          >
                            <div
                              class="MuiGrid-root MuiGrid-item"
                            >
                              <div
                                class="MuiBox-root MuiBox-root-25"
                              >
                                <div
                                  class="MuiFormControl-root MuiTextField-root CoTextField-root-18"
                                >
                                  <label
                                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                                    data-shrink="true"
                                  >
                                    Name
                                  </label>
                                  <div
                                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                                  >
                                    <input
                                      aria-invalid="false"
                                      class="MuiInputBase-input MuiInput-input"
                                      id="listFieldFormView.form.contacts[1].name"
                                      name="listFieldFormView.form.contacts[1].name"
                                      type="text"
                                      value="Jan Jansen"
                                    />
                                  </div>
                                </div>
                              </div>
                            </div>
                            <div
                              class="MuiGrid-root MuiGrid-item"
                            >
                              <div
                                class="MuiBox-root MuiBox-root-26"
                              >
                                <div
                                  class="MuiFormControl-root MuiTextField-root CoTextField-root-18"
                                >
                                  <label
                                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                                    data-shrink="true"
                                  >
                                    Email
                                  </label>
                                  <div
                                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                                  >
                                    <input
                                      aria-invalid="false"
                                      class="MuiInputBase-input MuiInput-input"
                                      id="listFieldFormView.form.contacts[1].email"
                                      name="listFieldFormView.form.contacts[1].email"
                                      type="text"
                                      value="jj@lv.nv"
                                    />
                                  </div>
                                </div>
                              </div>
                            </div>
                            <div
                              class="MuiGrid-root MuiGrid-item"
                            >
                              <div
                                class="MuiBox-root MuiBox-root-27"
                              >
                                <div
                                  class="MuiFormControl-root MuiTextField-root"
                                >
                                  <label
                                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                                    data-shrink="true"
                                  >
                                    Mobile
                                  </label>
                                  <div
                                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                                  >
                                    <input
                                      aria-invalid="false"
                                      class="MuiInputBase-input MuiInput-input"
                                      id="listFieldFormView.form.contacts[1].phone"
                                      name="listFieldFormView.form.contacts[1].phone"
                                      type="text"
                                      value="(702) 555-1212"
                                    />
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                    <button
                      class="MuiButtonBase-root MuiIconButton-root CoListField-closeMenuIcon-12"
                      tabindex="0"
                      type="button"
                    >
                      <span
                        class="MuiIconButton-label"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
                          />
                        </svg>
                      </span>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </li>
                </ul>
                <div
                  class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
                >
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-text"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root CoRowListFieldAddRowButton-leftIcon-28 MuiSvgIcon-fontSizeSmall"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                        />
                      </svg>
                      Add
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
